pkg/services/secrets/kvstore/remote_plugin.go

```diff
@@ -163,3 +168,45 @@ func updateFatalFlag(ctx context.Context, skv secretsKVStorePlugin) {
 func wrapUserFriendlySecretError(ufe string) datasources.ErrDatasourceSecretsPluginUserFriendly {
 	return datasources.ErrDatasourceSecretsPluginUserFriendly{Err: ufe}
 }
+
+func GetNamespacedKVStore(kv kvstore.KVStore) *kvstore.NamespacedKVStore {
+	return kvstore.WithNamespace(kv, kvstore.AllOrganizations, PluginNamespace)
+}
+
+func isPluginStartupErrorFatal(ctx context.Context, kvstore *kvstore.NamespacedKVStore) (bool, error) {
+	_, exists, err := kvstore.Get(ctx, QuitOnPluginStartupFailureKey)
+	if err != nil {
+		return false, errors.New(fmt.Sprint("error retrieving key ", QuitOnPluginStartupFailureKey, " from kvstore. error: ", err.Error()))
+	}
+	return exists, nil
+}
+
+func setPluginStartupErrorFatal(ctx context.Context, kvstore *kvstore.NamespacedKVStore, isFatal bool) error {
+	if !isFatal {
+		return kvstore.Del(ctx, QuitOnPluginStartupFailureKey)
+	}
+	return kvstore.Set(ctx, QuitOnPluginStartupFailureKey, "true")
+}
+
+func shouldUseRemoteSecretsPlugin(mg plugins.SecretsPluginManager, cfg *setting.Cfg) bool {
+	usePlugin := cfg.SectionWithEnvOverrides("secrets").Key("use_plugin").MustBool()
+	if !usePlugin {
+		logger.Debug("remote secret managements plugin disabled because the property `secrets.use_plugin` is not set to `true`")
+		return false
+	}
+	pluginInstalled := mg.SecretsManager() != nil
+	if !pluginInstalled {
+		logger.Debug("remote secret managements plugin disabled because there is no installed plugin of type `secretsmanager`")
+		return false
+	}
+	return true
+}
+
+func startAndReturnPlugin(mg plugins.SecretsPluginManager, ctx context.Context) (secretsmanagerplugin.SecretsManagerPlugin, error) {
```

Comment:
I do think it is still worth guarding against multiple calls to this function. I would suggest using sync.Once and sticking the start logic in there just in case.

Reply:
Will do!

Reply:
Done in 10b1f4fea7